Justin_Output-section-CRUD_1203077277702737

pauljustinhernandez opened the pull request:
ASANA ticket: https://app.asana.com/0/1203106518351809/1203077277702737/f
Initial PR content reference: https://app.asana.com/0/0/1203077277702737/1203130611238452/f

Description: Amended edit output function to suffice the logics for the other output types as per Tian

darlp5589 reviewed (changes requested):
Done checking. Here are my concerns:

1. I am adding a newly but uncommitted Transport w/ ATSC 3 type, it's not being pulled in the output.
2. i1: Please change the title to Add Output based on client app. Default output type in dropdown is ASI.
3. v1: Selecting an ATSC 3 Transport and clicked Close (instead of Select) button, it will still select that Transport.
4. i2: Viewing a transport should display the Transport name in the Title Bar, not the TSID value and it says BSID, not TSID from the Client App.
5. i3: Please include 0x5 hexacode on TSID part.
6. v2: I have selected a Transport then removed it (cleared already) and clicked Prev button. When I clicked next button, it still retain the value of the supposedly cleared Transport.
7. v3: I have added a Transport, then removed and clicked Add. It still saved the supposedly removed Transport.
8. i4: What is the difference between 0 and a dash transport? In Client App, they are both 0.
9. v4: I have created 2 outputs: out1ab and out2. When I attempted to update out2, it will point out o out1ab.
10. v5: A scenario wherein I have 2 outputs: out2 and out1a. When I edit out1a to be out2, it will display an error saying it exists which is correct thus DISABLING the NEXT button. When I close it and clicked edit again, that next button is still DISABLE and I can't edit to add a transport because I have to make some changes on the output name.
11. v6: I have added output w/o a transport yet then updated that output and added a transport then clicked remove that transport and save. It will still save that supposedly removed transport. This will also reflect in the big commit part.
12. I have added an output with a transport and saved. When I renamed that transport in the transport tab, the transport being linked to an output isn't renamed but pointing to the old name.
13. In relation to concern 12, it is the same case with manually deleting a transport.


i1: 
![image](https://user-images.githubusercontent.com/101298313/195212856-8493a7fd-448e-4046-8ab6-145a7e7ff324.png)

v1: https://www.loom.com/share/fca9e1d98fee4a66833fe09a802107da

i2: 
![image](https://user-images.githubusercontent.com/101298313/195214414-f424cadb-231d-4979-b8af-f964ee46896c.png)

i3: 
![image](https://user-images.githubusercontent.com/101298313/195214792-53f591bd-f75b-42c8-87bc-f8a3d1206c64.png)

v2: https://www.loom.com/share/f1970346da9243358970e0167762cb20

v3: https://www.loom.com/share/8fa93559be8f48679a796174ebcd9ae8

i4: 
![image](https://user-images.githubusercontent.com/101298313/195220227-e6982903-ad69-42ab-b34b-2b8f169f32c3.png)

v4: https://www.loom.com/share/7fb375ac46a447ab8dc38640a31706a5

v5: https://www.loom.com/share/8cac186528fe4d75a5a2652c40000b72

v6: https://www.loom.com/share/85a7e4c59cfe4173a0b0d4902fc25db9

darlp5589 commented on src/app/pages/dashboard/network-config/output/output.component.ts:
```diff
@@ -319,23 +319,25 @@ export class OutputComponent
-              } else {
-                const editAdd = this.addData;
+
+              let dataRow = tData.find((obj) => obj.id === tableID);
```
optional chaining

darlp5589 commented on src/app/pages/dashboard/network-config/output/output.component.ts:
```diff
@@ -319,23 +319,25 @@ export class OutputComponent
-                  (obj) => obj.addIdx === tableIDx
-                );
-                this.btnEditOutputFunction(dataRowedit, 'addedit');
+                dataRow = tData.find((obj) => obj.clientId === tableIDx);
```
optional chaining

darlp5589 reviewed (changes requested):
Done checking half of the fixes and confirmed fine on most of them but noticed another concern.

1. i1: Kindly remove the hexacode after removing the transport.
2. v1: I have added an output, linked to a transport then viewed it, closed the viewed transport and clicked prev. The output information was gone.
3. v2: There's a really random weird instance quite hard to replicate but at some point along testing, after removing a transport and linking back again, I can't see anymore available transport display with a console error. Sometimes, it occurred early, sometimes at the later part of testing.
4. v3: I have added an output w/o a transport at first then edited and added a transport on it then saved. Now when I edited it and removed that transport and clicked update, a console error appeared and I can't proceed on the update.
5. i2: I still got a combination of output list with a 0 and - under transport column.


i1: 
![image](https://user-images.githubusercontent.com/101298313/195471196-60395af7-e662-4dfb-8b9c-44ea22320286.png)

v1: https://www.loom.com/share/4a3495f2fc2947879961181710f24644

v2: https://www.loom.com/share/35ac8b0f971843e8876228b68b6fa80f

v3: https://www.loom.com/share/8a2e44cd21b743c7bef3e6b9e6f37e34

i2: 
![image](https://user-images.githubusercontent.com/101298313/195473490-ae2bbc24-89e5-455a-8361-8e85719f7394.png)

pauljustinhernandez commented:
> 9. v4: I have created 2 outputs: out1ab and out2. When I attempted to update out2, it will point out o out1ab.
**- OK. Ready for testing**

> 10. v5: A scenario wherein I have 2 outputs: out2 and out1a. When I edit out1a to be out2, it will display an error saying it exists which is correct thus DISABLING the NEXT button. When I close it and clicked edit again, that next button is still DISABLE and I can't edit to add a transport because I have to make some changes on the output name.
**- OK. Ready for testing**

> 11. v6: I have added output w/o a transport yet then updated that output and added a transport then clicked remove that transport and save. It will still save that supposedly removed transport. This will also reflect in the big commit part.
**- OK. Ready for testing**


**As discussed, the ff. items are included in the out of scope ticket**

> 12. I have added an output with a transport and saved. When I renamed that transport in the transport tab, the transport being linked to an output isn't renamed but pointing to the old name.
> 13. In relation to concern 12, it is the same case with manually deleting a transport.
> 1. I am adding a newly but uncommitted Transport w/ ATSC 3 type, it's not being pulled in the output.

cc: @darlp5589

darlp5589 reviewed (changes requested):
Done checking fixes. Here are my concerns:

1. i1: Added an output then edited the output with a modal title "Edit Output - out1" and clicked next. When I clicked on prev button, the modal title should revert to "Edit Output - out1".
2. i2: Added an output (out1a) then edited that output (out1ab) and clicked next. The modal title says "Edit ATSC 3.0 UDP Output out1ab". It should still point to the original name since it's not saved yet based on the client app. Also add a dash in between. So it should just say "Edit ATSC 3.0 UDP Output - out1a".
3. i3: Comparing the same list of outputs from Java App and Angular App, kindly make sure Java app is using 1 on outputs with transports and last update column is in date values.



i1: 
![image](https://user-images.githubusercontent.com/101298313/195738176-04e777eb-ef71-4cd1-8b35-7d89316c8735.png)

i2: 
![image](https://user-images.githubusercontent.com/101298313/195738588-7d6b0a5e-e944-47db-9019-c8ab78e9f9a2.png)

i3: 
![image](https://user-images.githubusercontent.com/101298313/195739792-7dc857cd-1cb2-47e7-8504-6d2f92f2ba23.png)

pauljustinhernandez commented:
Hi Darl, some of the issues were fixed in the last commit.

**READY FOR TESTING**
1. Kindly remove the hexacode after removing the transport.
2. I still got a combination of output list with a 0 and - under transport column.
3. Added an output then edited the output with a modal title "Edit Output - out1" and clicked next. When I clicked on prev button, the modal title should revert to "Edit Output - out1".
4. Added an output (out1a) then edited that output (out1ab) and clicked next. The modal title says "Edit ATSC 3.0 UDP Output out1ab". It should still point to the original name since it's not saved yet based on the client app. Also add a dash in between. So it should just say "Edit ATSC 3.0 UDP Output - out1a".


**TO DO**
1. I have added an output, linked to a transport then viewed it, closed the viewed transport and clicked prev. The output information was gone.
2. There's a really random weird instance quite hard to replicate but at some point along testing, after removing a transport and linking back again, I can't see anymore available transport display with a console error. Sometimes, it occurred early, sometimes at the later part of testing.
3. I have added an output w/o a transport at first then edited and added a transport on it then saved. Now when I edited it and removed that transport and clicked update, a console error appeared and I can't proceed on the update.
4. Comparing the same list of outputs from Java App and Angular App, kindly make sure Java app is using 1 on outputs with transports and last update column is in date values.

CC: @darlp5589

darlp5589 reviewed (changes requested):
Done checking on fixes. I have another findings below though:

1.v1: I have an output from server, edited it and removed the linked transport then clicked prev, the output type is gone and when you clicked next, the whole modal is empty.
2. v2: Though I have logged this already and it's still under TO DO items, previously I can't barely replicate the scenario but during my recent testing, I was able to. To do so, create an output, linked to a transport, click [prev button, remove the linked transport and re-linked again, no transport details will be shown or the modal is empty.
3. v3: I still got the combination of 0 and - in transport column. This time I have existing set of outputs from server then added 2 outputs w/o linked transports then edited the 2nd added transport w/o doing anything and clicked update. 



v1: https://www.loom.com/share/3f789e5160614bda9ea0a2a109d21580
v2: https://www.loom.com/share/5bb763adb06c4e1ea834956e37e73e57
v3: https://www.loom.com/share/41619aadacb14cf9bbbd3da0fc91a56c

pauljustinhernandez commented:
Hi Darl, some of the issues were fixed in the last commit.

**READY FOR TESTING**
1. I have added an output, linked to a transport then viewed it, closed the viewed transport and clicked prev. The output information was gone.

2. I have an output from server, edited it and removed the linked transport then clicked prev, the output type is gone and when you clicked next, the whole modal is empty.

3. Though I have logged this already and it's still under TO DO items, previously I can't barely replicate the scenario but during my recent testing, I was able to. To do so, create an output, linked to a transport, click [prev button, remove the linked transport and re-linked again, no transport details will be shown or the modal is empty.

4. There's a really random weird instance quite hard to replicate but at some point along testing, after removing a transport and linking back again, I can't see anymore available transport display with a console error. Sometimes, it occurred early, sometimes at the later part of testing.

5. I have added an output w/o a transport at first then edited and added a transport on it then saved. Now when I edited it and removed that transport and clicked update, a console error appeared and I can't proceed on the update.


**TO DO**
1. Comparing the same list of outputs from Java App and Angular App, kindly make sure Java app is using 1 on outputs with transports and last update column is in date values.

2. I still got the combination of 0 and - in transport column. This time I have existing set of outputs from server then added 2 outputs w/o linked transports then edited the 2nd added transport w/o doing anything and clicked update.

cc: @darlp5589

darlp5589 reviewed (changes requested):
Done checking fixes. Here are my new set of concerns:

1.v1: I have an existing output from server w/ linked transport, edited it then removed the linked transport and clicked update button. In the output list table, it's already marked as 0 under transport column but if you edit, it's still linking to the transport.
2. v2: I have an existing output from server w/o a linked transport. Edited it and added a transport then clicked close or cancel. When I edited it again, it keeps linking to the transport even if it was cancelled.


v1: https://www.loom.com/share/6e5629f68ac746ff9286ddd6048af1a0
v2: https://www.loom.com/share/5676a7fdf4b1418991675eef4cf25eb7

pauljustinhernandez commented:
Hi Darl, some of the issues were fixed in the last commit.

**READY FOR TESTING**
1. Comparing the same list of outputs from Java App and Angular App, kindly make sure Java app is using 1 on outputs with transports and last update column is in date values. **(LAST UPDATE not included, see TO DO list below)**

2. I still got the combination of 0 and - in transport column. This time I have existing set of outputs from server then added 2 outputs w/o linked transports then edited the 2nd added transport w/o doing anything and clicked update.


**TO DO**
1. Last Update column, there's no available data for this. need to discussed with the team

2. I have an existing output from server w/ linked transport, edited it then removed the linked transport and clicked update button. In the output list table, it's already marked as 0 under transport column but if you edit, it's still linking to the transport. cc: @chriscodev , Tian helping me to fix this issue.

3. I have an existing output from server w/o a linked transport. Edited it and added a transport then clicked close or cancel. When I edited it again, it keeps linking to the transport even if it was cancelled. (correlates to the #2 issue) cc: @chriscodev , Tian helping me to fix this issue.


cc: @darlp5589

darlp5589 reviewed (changes requested):
Done checking on fixes. I still got the - (dash) value under transport column. I have 4 existing outputs from server and added 1 output with/without linked transport (out5). It should be 1 or 0 instead of - (dash).

![image](https://user-images.githubusercontent.com/101298313/196614299-034d355f-6a70-4974-b15f-bb244d3f40e6.png)

pauljustinhernandez commented:
Hi Darl, some of the issues were fixed in the last commit.

**READY FOR TESTING**
1. I have an existing output from server w/o a linked transport. Edited it and added a transport then clicked close or cancel. When I edited it again, it keeps linking to the transport even if it was cancelled.

2. Done checking on fixes. I still got the - (dash) value under transport column. I have 4 existing outputs from server and added 1 output with/without linked transport (out5). It should be 1 or 0 instead of - (dash).


**TO DO**
1. Last Update column, there's no available data for this. need to discussed with the team

2. I have an existing output from server w/ linked transport, edited it then removed the linked transport and clicked update button. In the output list table, it's already marked as 0 under transport column but if you edit, it's still linking to the transport. 


cc: @darlp5589

darlp5589 reviewed (changes requested):
Done checking fixes and they are good but I noticed another concerns. Not sure if these are branching or regressed issue but you may verify them:

1. v1: I have an existing output w/o a linked transport from server, updated it and linked a transport then clicked update. In the output list, it's already transport=1 but if you edit it again, there's no linked transport even if it's updated already.
2. v2: This is an opposite scenario of your pending TO DO (2nd item). So I have an output from server with a linked transport, updated it and removed the linked transport then clicked CANCEL (it should still link). In the output list, it's still transport=1 which is correct but if you edit it again, there's no anymore linked transport.
3. v3: My previous second concern still exists on both with or without a linked transport. (I still got the - (dash) value under transport column. I have 4 existing outputs from server and added 1 output with/without linked transport (out5). It should be 1 or 0 instead of - (dash).)
4. v3: Added output 5 with linked transport on top of existing 4 outputs from server. When I edited it and removed the linked transport then CANCEL instead of UPDATE, numerous console.log errors are throwing and it's running endlessly.


v1: https://www.loom.com/share/4115498c32c04d6390f8fb9d91cffed8
v2: https://www.loom.com/share/cea9a095141a4806a3dcd3f80c543f3b
v3: https://www.loom.com/share/db5f25e0218240ec92c584bc588e1283
